workers/unity/Packages/io.improbable.gdk.core/Systems/EntitySystem.cs

```diff
@@ -1,61 +1,127 @@
                 }
 
-                if (entitiesAdded.Count != 0 || entitiesRemoved.Count != 0)
+                EntityIds.Dispose();
+                EntityIds = new NativeArray<EntityId>(EntityCount << 1, Allocator.Persistent);
+            }
+
+            public void Dispose()
+            {
+                if (EntityIds.IsCreated)
                 {
-                    ViewVersion += 1;
+                    EntityIds.Dispose();
                 }
             }
         }
 
+        private EndSimulationEntityCommandBufferSystem bufferSystem;
+        private EntityCollection added;
+        private EntityQuery addedQuery; // This query needs to be a member of SystemBase
+        private EntityCollection removed;
+        private EntityQuery removedQuery; // This query needs to be a member of SystemBase
+        public int ViewVersion { get; private set; }
+        public NativeSlice<EntityId> EntitiesRemoved => removed.Slice;
+        public NativeSlice<EntityId> EntitiesAdded => added.Slice;
+
+        protected override void OnCreate()
+        {
+            base.OnCreate();
+            bufferSystem = World.GetOrCreateSystem<EndSimulationEntityCommandBufferSystem>();
+            removed = new EntityCollection(1);
+            added = new EntityCollection(1);
+        }
+
         protected override void OnUpdate()
         {
-            entitiesAdded.Clear();
-            entitiesRemoved.Clear();
+            UpdateEntitiesAdded();
+            UpdateEntitiesRemoved();
+
+            Dependency = JobHandle.CombineDependencies(added.JobHandle, removed.JobHandle);
+            bufferSystem.AddJobHandleForProducer(Dependency);
+
+            if (added.EntityCount != 0 || removed.EntityCount != 0)
+            {
+                ViewVersion += 1;
+            }
+        }
+
+        private void UpdateEntitiesAdded()
+        {
+            added.EnsureSize(ref addedQuery);
+            var buffer = bufferSystem.CreateCommandBuffer().AsParallelWriter();
+            var array = added.EntityIds;
+            added.JobHandle = Entities.WithName("EntitiesAdded")
+                .WithNone<EntitySystemStateComponent>()
+                .WithAll<SpatialEntityId>()
+                .WithStoreEntityQueryInField(ref addedQuery)
+                .ForEach((int entityInQueryIndex, in Entity entity, in SpatialEntityId entityId) =>
+                {
+                    buffer.AddComponent(entityInQueryIndex, entity, (EntitySystemStateComponent) entityId);
+                    array[entityInQueryIndex] = entityId.EntityId;
+                }).ScheduleParallel(Dependency);
+        }
+
+        private void UpdateEntitiesRemoved()
+        {
+            removed.EnsureSize(ref removedQuery);
+            var buffer = bufferSystem.CreateCommandBuffer().AsParallelWriter();
+            var array = removed.EntityIds;
+            removed.JobHandle = Entities.WithName("EntitiesRemoved")
+                .WithAll<EntitySystemStateComponent>()
```

Comment:
No need for this, as this constraint is already created with the `in EntitySystemStateComponent entityId`